[GTK][WPE] Add new JavaScript execution APIs

carlosgcampos opened the pull request:
#### c07c7ce77b199a76905c2adeff5f4cfc2cc06a33
<pre>
[GTK][WPE] Add new JavaScript execution APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=241420">https://bugs.webkit.org/show_bug.cgi?id=241420</a>

Reviewed by Adrian Perez de Castro and Michael Catanzaro.

Add webkit_web_view_evaluate_javascript() and
webkit_web_view_call_async_javascript_function() and deprecate all other
JavaScript execution funtions. The new functions add a length parameter
to allow running code containing null characters and the source URI that
is shown in exception messages.

* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
(webkitWebViewRunJavascriptWithoutForcedUserGestures):
(webkit_web_view_evaluate_javascript):
(webkit_web_view_evaluate_javascript_finish):
(parseAsyncFunctionArguments):
(webkit_web_view_call_async_javascript_function):
(webkit_web_view_call_async_javascript_function_finish):
(webkit_web_view_run_javascript):
(webkit_web_view_run_javascript_in_world):
(webkit_web_view_run_async_javascript_function_in_world):
(resourcesStreamReadCallback):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::updateTargetList):
* Tools/MiniBrowser/gtk/BrowserTab.c:
(browser_tab_load_uri):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestBackForwardList.cpp:
(testWebKitWebViewSessionStateWithFormData):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestResources.cpp:
(testWebViewSyncRequestOnMaxConns):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestSSL.cpp:
(WebSocketTest::connectToServerAndWaitForEvents):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebExtensions.cpp:
(FormSubmissionTest::runJavaScriptAndWaitUntilFormSubmitted):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
(isStyleSheetInjectedForURLAtPath):
(isScriptInjectedForURLAtPath):
(UserScriptMessageTest::postMessageAndWaitUntilReceived):
(UserScriptMessageTest::runAsyncJavaScriptFinished):
(UserScriptMessageTest::postMessageAndWaitForPromiseResolved):
(testUserContentManagerScriptMessageInWorldReceived):
(testUserContentManagerScriptMessageWithReplyReceived):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:
(ProxyTest::createWebSocketAndWaitUntilConnected):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(testWebViewRunJavaScript):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.cpp:
(runJavaScriptReadyCallback):
(runAsyncJavaScriptFunctionInWorldReadyCallback):
(WebViewTest::runJavaScriptAndWaitUntilFinished):
(WebViewTest::runJavaScriptFromGResourceAndWaitUntilFinished):
(WebViewTest::runJavaScriptInWorldAndWaitUntilFinished):
(WebViewTest::runAsyncJavaScriptFunctionInWorldAndWaitUntilFinished):
(WebViewTest::runJavaScriptWithoutForcedUserGesturesAndWaitUntilFinished):
(WebViewTest::runJavaScriptAndWait):
(runJavaScriptFromGResourceReadyCallback): Deleted.
(runJavaScriptInWorldReadyCallback): Deleted.
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.h:

Canonical link: <a href="https://commits.webkit.org/259852@main">https://commits.webkit.org/259852@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a15aabfbd96fcc6f95ac10583fd6ac6b713840d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106017 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38850 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115199 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175353 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109920 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6253 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98226 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114934 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12559 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95529 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40101 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94424 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27172 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81777 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8327 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28524 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8817 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5081 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14441 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48067 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10361 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3668 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -136,11 +136,15 @@ typedef enum {
  */
 typedef enum {
-    WEBKIT_JAVASCRIPT_ERROR_SCRIPT_FAILED = 699
+    WEBKIT_JAVASCRIPT_ERROR_SCRIPT_FAILED = 699,
+    WEBKIT_JAVASCRIPT_ERROR_INVALID_PARAMETER = 600,
+    WEBKIT_JAVASCRIPT_ERROR_INVALID_RESULT = 601,
```
Any particular reason for the selected values? It's fine, just seems like strange choices.

carlosgcampos replied:
That's what we do for the other errors, but I don't remember why. I think some of the values are shared with WebCore.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3915,55 +3917,285 @@ JSGlobalContextRef webkit_web_view_get_javascript_global_context(WebKitWebView*
+                if (exceptionDetails.lineNumber > 0)
+                    builder.append(':', exceptionDetails.lineNumber);
+                if (exceptionDetails.columnNumber > 0)
+                    builder.append(':', exceptionDetails.columnNumber);
```
I know this should surely never happen, but if somehow we wind up with a column number but no line number, then this will print something very confusing, so better:

```suggestion
                if (exceptionDetails.lineNumber > 0) {
                    builder.append(':', exceptionDetails.lineNumber);
                    if (exceptionDetails.columnNumber > 0)
                        builder.append(':', exceptionDetails.columnNumber);
                }
```

carlosgcampos replied:
Good point

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3915,55 +3917,285 @@ JSGlobalContextRef webkit_web_view_get_javascript_global_context(WebKitWebView*
+ * @user_data: (closure): the data to pass to callback function
+ *
+ * Asynchronously evaluate @script in the script world with name @world_name of the main frame current context in @web_view.
+ * If @world_name is %NULL the default world is used. Any value that is not %NULL is a distinct world.
```
```suggestion
 * If @world_name is %NULL, the default world is used. Any value that is not %NULL is a distinct world.
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3915,55 +3917,285 @@ JSGlobalContextRef webkit_web_view_get_javascript_global_context(WebKitWebView*
+ * Asynchronously evaluate @script in the script world with name @world_name of the main frame current context in @web_view.
+ * If @world_name is %NULL the default world is used. Any value that is not %NULL is a distinct world.
+ * The @source_uri will be shown in exceptions and doesn't affect the behavior of the script.
+ * When not provided the document URL is used.
```
```suggestion
 * When not provided, the document URL is used.
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3915,55 +3917,285 @@ JSGlobalContextRef webkit_web_view_get_javascript_global_context(WebKitWebView*
+ * The @source_uri will be shown in exceptions and doesn't affect the behavior of the script.
+ * When not provided the document URL is used.
+ *
+ * Note that if #WebKitSettings:enable-javascript is %FALSE, this method will do nothing.
```
```suggestion
 * Note that if #WebKitSettings:enable-javascript is %FALSE, this method will do nothing.
 * If you want to use this method but still prevent web content from executing its own
 * JavaScript, then use #WebKitSettings:enable-javascript-markup.
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3915,55 +3917,285 @@ JSGlobalContextRef webkit_web_view_get_javascript_global_context(WebKitWebView*
+ */
+WebKitJavascriptResult* webkit_web_view_evaluate_javascript_finish(WebKitWebView* webView, GAsyncResult* result, GError** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_WEB_VIEW(webView), nullptr);
+    g_return_val_if_fail(g_task_is_valid(result, webView), nullptr);
```
There's another safety check that would be good to start doing:

```
g_return_val_if_fail(g_task_get_source_tag(task) == webkit_web_view_evaluate_javascript, nullptr);
```

For this to work, you'd need to call g_task_set_source_tag(), which will make the code slightly more complicated since you won't be able to create the task on the same line that you pass it to `webkitWebViewRunJavaScriptWithParams` anymore, but the extra check is worth it IMO.

carlosgcampos replied:
We don't use tags in any of the async apis, if we decide to use them we should do it for all other cases in a separate patch, I think.

mcatanzaro replied:
I will add this as a TODO for myself, then.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -3915,55 +3917,285 @@ JSGlobalContextRef webkit_web_view_get_javascript_global_context(WebKitWebView*
+ */
+WebKitJavascriptResult* webkit_web_view_call_async_javascript_function_finish(WebKitWebView* webView, GAsyncResult* result, GError** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_WEB_VIEW(webView), nullptr);
+    g_return_val_if_fail(g_task_is_valid(result, webView), nullptr);
```
Ditto.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -4160,47 +4382,7 @@ void webkit_web_view_run_javascript_in_world(WebKitWebView* webView, const gchar
  */
 void webkit_web_view_run_async_javascript_function_in_world(WebKitWebView* webView, const gchar* body, GVariant* arguments, const char* worldName, GCancellable* cancellable, GAsyncReadyCallback callback, gpointer userData)
```
You forgot to mark this one as deprecated in the doc comment.

carlosgcampos replied:
Good catch

carlosgcampos committed:
[GTK][WPE] Add new JavaScript execution APIs
https://bugs.webkit.org/show_bug.cgi?id=241420

Reviewed by Adrian Perez de Castro and Michael Catanzaro.

Add webkit_web_view_evaluate_javascript() and
webkit_web_view_call_async_javascript_function() and deprecate all other
JavaScript execution funtions. The new functions add a length parameter
to allow running code containing null characters and the source URI that
is shown in exception messages.

* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
(webkitWebViewRunJavascriptWithoutForcedUserGestures):
(webkit_web_view_evaluate_javascript):
(webkit_web_view_evaluate_javascript_finish):
(parseAsyncFunctionArguments):
(webkit_web_view_call_async_javascript_function):
(webkit_web_view_call_async_javascript_function_finish):
(webkit_web_view_run_javascript):
(webkit_web_view_run_javascript_in_world):
(webkit_web_view_run_async_javascript_function_in_world):
(resourcesStreamReadCallback):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::updateTargetList):
* Tools/MiniBrowser/gtk/BrowserTab.c:
(browser_tab_load_uri):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestBackForwardList.cpp:
(testWebKitWebViewSessionStateWithFormData):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestResources.cpp:
(testWebViewSyncRequestOnMaxConns):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestSSL.cpp:
(WebSocketTest::connectToServerAndWaitForEvents):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebExtensions.cpp:
(FormSubmissionTest::runJavaScriptAndWaitUntilFormSubmitted):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
(isStyleSheetInjectedForURLAtPath):
(isScriptInjectedForURLAtPath):
(UserScriptMessageTest::postMessageAndWaitUntilReceived):
(UserScriptMessageTest::runAsyncJavaScriptFinished):
(UserScriptMessageTest::postMessageAndWaitForPromiseResolved):
(testUserContentManagerScriptMessageInWorldReceived):
(testUserContentManagerScriptMessageWithReplyReceived):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:
(ProxyTest::createWebSocketAndWaitUntilConnected):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(testWebViewRunJavaScript):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.cpp:
(runJavaScriptReadyCallback):
(runAsyncJavaScriptFunctionInWorldReadyCallback):
(WebViewTest::runJavaScriptAndWaitUntilFinished):
(WebViewTest::runJavaScriptFromGResourceAndWaitUntilFinished):
(WebViewTest::runJavaScriptInWorldAndWaitUntilFinished):
(WebViewTest::runAsyncJavaScriptFunctionInWorldAndWaitUntilFinished):
(WebViewTest::runJavaScriptWithoutForcedUserGesturesAndWaitUntilFinished):
(WebViewTest::runJavaScriptAndWait):
(runJavaScriptFromGResourceReadyCallback): Deleted.
(runJavaScriptInWorldReadyCallback): Deleted.
* Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.h:

Canonical link: https://commits.webkit.org/259852@main
